### ISSUE TITLE
Prisma Client Edge: environment variables are not working with the "new" Module Worker syntax for Cloudflare Workers

### ISSUE THREAD

**Issue by Abiti-936** [prisma/prisma#15958]
## Problem

According to the recent changes in Cloudflare Workers (https://developers.cloudflare.com/workers/runtime-apis/fetch-event/) & Wrangler CLI, Environment Variables access Model, environment variables aren't globally available anymore & in the generated edge client a Database Url is accessed through process.env.DATABASE_URL (which workers on Vercel Edge Functions) or using global. I think the Prisma Client should accept Database Url when it's instantiated, because Different Edge Platform have their own way of accessing environment variables for example, in Netlify Edge Functions, it's like

```ts
const db_url = Deno.env.get("DATABASE_URL");
```
and in Cloudflare Workers, it's like 

```ts
export default {  
  fetch(request, env, context) {
    const db_url = env.DATABASE_URL;
  },
};

```

or in fastly compute edge using dictionaries, it's like

```ts
  const tokens = new Dictionary("tokens");
  const db_url = tokens.get("DATABASE_URL");
```

## Suggested solution

I think Prisma Client should accept a datebase url when it's instantiated like

```ts
import { PrismaClient } from '@prisma/client'

const prisma = new PrismaClient({
  url: "" // Pass Database Url
});
```

**Comment by janpio** [comment]
This already exists: https://www.prisma.io/docs/reference/api-reference/prisma-client-reference#programmatically-override-a-datasource-url

But Prisma should also always be able to read a value set in an environment variable, e.g. via `url = env("DATABASE_URL")` in the Prisma Schema. Is that currently not the case?

**Comment by Abiti-936** [comment]
Yes, that works but, My Main Problem is in the generated edge client Code with the client config. I have to replace these database url environment variables all check code to use

```ts
import { PrismaClient } from '@prisma/client'

const prisma = new PrismaClient({
  datasources: {
    db: {
      url: env.DATABASE_URL
    },
  },
})
```

![Screenshot_20221024-212014_Brave](https://user-images.githubusercontent.com/103219498/197598069-ece3b694-9043-4c3f-8406-8fa23fbba861.jpg)

**Comment by janpio** [comment]
I am not following. When exactly is what failing why? Please use a concrete example so this is as concrete as possible. Thanks.

**Comment by Abiti-936** [comment]
I have been using Cloudflare Workers & the new Workers Module Syntax exposes environment variables with **env** arguments passed to the fetch function & accessing environment variables globally is going to fail.

```ts
export default {  
  fetch(request, env, context) { 
    console.log(DATABASE_URL) // is going to fail
    console.log(env.DATABASE_URL) // works
  },
};
```

**Comment by janpio** [comment]
If `process.env.DATABASE_URL` also does not work any more, that is indeed a problem for our default way of doing things.

**Comment by Abiti-936** [comment]
I think you can leave it the default typeof "process" !== undefined && process.env.DATABASE_URL unchanged & let people pass the database url manually. Thanks.

**Comment by janpio** [comment]
We want things to work out of the box optimally - because otherwise, you will need to have read the documentation which might or might not be the case.

We will treat this as a bug.

**Comment by Jolg42** [comment]
Indeed, there is now a new syntax from Cloudflare since 16/11/2021 https://blog.cloudflare.com/workers-javascript-modules/

Both are supported.

Old syntax = [Service Worker](https://developers.cloudflare.com/workers/runtime-apis/fetch-event/#syntax-service-worker)
>When a Worker is deployed using the Service Worker syntax, any [bindings](https://developers.cloudflare.com/workers/platform/environment-variables/) will be made available as global runtime variables.
```
addEventListener('fetch', event => {
  event.respondWith(new Response('Hello'));
});
```

New Syntax = [Module Worker](https://developers.cloudflare.com/workers/runtime-apis/fetch-event/#syntax-module-worker)
>When deploying a Module Worker, any [bindings](https://developers.cloudflare.com/workers/platform/environment-variables/) will not be available as global runtime variables. Instead, they are passed to the handler as a [parameter](https://developers.cloudflare.com/workers/runtime-apis/fetch-event/#parameters) – refer to env in [Parameters](https://developers.cloudflare.com/workers/runtime-apis/fetch-event/#parameters).
```
export default {
  fetch(request, env, context) {
    return new Response('Hello');
  },
};
```

Benefits from https://developers.cloudflare.com/workers/learning/migrating-to-module-workers/
- [Durable Objects](https://developers.cloudflare.com/workers/learning/using-durable-objects/) require the module syntax.
- Module Workers do not rely on any global bindings, which means the Workers runtime does not need to set up fresh execution contexts, making Module Workers safer and faster to run.
- Module Workers are ES Modules, which allows them to be shared and published to npm, for example. Module Workers can be imported by and composed within other Module Workers.

**Comment by Abiti-936** [comment]
I think someone who's starting a new Cloudflare Workers Project with Wrangler CLI 2 will use the (Module Syntax)[https://developers.cloudflare.com/workers/runtime-apis/fetch-event/#syntax-module-worker], because wrangler 2 generates a project with a Module Syntax by default, people will see many applications crashes, so please update the environment variables default accessing method. Thanks.

**Comment by Abiti-936** [comment]
I think someone who's starting a new Cloudflare Workers Project with Wrangler CLI 2 will use the (Module Syntax)[https://developers.cloudflare.com/workers/runtime-apis/fetch-event/#syntax-module-worker], because wrangler 2 generates a project with a Module Syntax by default, people will see many applications crashes, so please update the environment variables default accessing method. Thanks.

**Comment by Jolg42** [comment]
Note: we could do a reproduction in prisma/ecosystem-tests
Based on
https://github.com/prisma/ecosystem-tests/tree/dev/dataproxy/cloudflare-workers
But using wrangler v2 with Module syntax

**Comment by inssein** [comment]
Just ran into this myself as well, a pretty broken experience when trying to use Remix with Cloudflare Pages and Prisma.

The only way I got around it (inside a resource route) was to set a bad database url in `schema.prisma` and pass one in myself when creating the PrismaClient.

Still stuck on server rendered pages.

**Comment by Jolg42** [comment]
Note: the docs page https://www.prisma.io/docs/guides/deployment/deployment-guides/deploying-to-cloudflare-workers will be updated to use wrangler v2 once https://github.com/prisma/docs/pull/4014 is merged

**Comment by ticup** [comment]
Ran into same issue with trying to run Prisma with Prisma Data Proxy on the [Netlify Edge Functions](https://docs.netlify.com/edge-functions/overview/), where they expose environment variables through [Deno.env](https://docs.netlify.com/edge-functions/api/#deno-apis), as pointed out by @Abiti-936.

This is the piece for me to get Remix + Prisma to run on the Edge Functions.

**Comment by Abiti-936** [comment]
I think for Netlify Edge Functions & Deno Deploy, there is a preview flag that you can set in your schema.prisma :-

```prisma
generator client {
  provider = "prisma-client-js"
  previewFeatures = ["deno"]
  output = "../generated/client"
}

datasource db {
  provider = "postgresql"
  url = env("DATABASE_URL")
}

```

as explained in [Prisma docs](https://www.prisma.io/docs/guides/deployment/deployment-guides/deploying-to-deno-deploy#9-deploy-to-deno-deploy)

**Comment by Abiti-936** [comment]
I think for Netlify Edge Functions & Deno Deploy, there is a preview flag that you can set in your schema.prisma :-

```prisma
generator client {
  provider = "prisma-client-js"
  previewFeatures = ["deno"]
  output = "../generated/client"
}

datasource db {
  provider = "postgresql"
  url = env("DATABASE_URL")
}

```

as explained in [Prisma docs](https://www.prisma.io/docs/guides/deployment/deployment-guides/deploying-to-deno-deploy#9-deploy-to-deno-deploy)

**Comment by ticup** [comment]
Thanks for the pointer.
Already had that set up and it doesn't help.

Must say that I'm running this with Remix as well, so problem might also lay there.

**Comment by greym0uth** [comment]
@janpio Would a viable option for cloudflare be requiring the datasource to be setup manually?

```ts
// For new module syntax
export default {
  async fetch(request, env) {
    const prisma = new PrismaClient({
      datasources: {
        db: {
          url: env.DATABASE_URL,
        },
      },
    });
    const users = await prisma.user.findMany();
    return Response.json(users);
  },
}

// For service worker syntax
const prisma = new PrismaClient({
  datasources: {
    db: {
      url: env.DATABASE_URL,
    },
  },
});
addEventListener('fetch', (event) => {
  const users = await prisma.user.findMany();
  event.respondWith(Response.json(users);
});
```

If so, this would be as simple as removing [these lines](https://github.com/prisma/prisma/blob/b70f3a52fedb9b36cac6031e2e56d13c8d70f1b9/packages/client/src/generation/utils/buildInjectableEdgeEnv.ts#L73-L74) from the generated output.

**Comment by janpio** [comment]
We ran into a related problem ourselves internally with Cloudflare pages: https://github.com/prisma/prisma/issues/18204 After an investigation of that we came to the exact same conclusion as @inssein:

> The only way I got around it (inside a resource route) was to set a bad database url in `schema.prisma` and pass one in myself when creating the PrismaClient.

So:
- Use `env(FOO)` in the schema file. 
- Use a `datasources.db.url` override with the correct `env.DATABASE_URL` in the code that does `new PrismaClient`, as in the example @greym0uth shared under "new module syntax".

**Comment by janpio** [comment]
@inssein YOu wrote "Still stuck on server rendered pages.". Can you elaborate on that a bit, maybe do a minimal reproduction?

**Comment by janpio** [comment]
https://github.com/prisma/prisma/issues/18204 was a duplicate of this issue, so I closed it, but included a full reproduction and the error code snippet:
 ```
 [pages:err] ReferenceError: DATABASE_URL_PROXY is not defined.
Attempted to access binding using global in modules.
You must use the 2nd `env` parameter passed to exported handlers/Durable Object constructors, or `context.env` with Pages Functions.
    at EventTarget.get (/workspace/repro-prisma-client-cloud-flare-pages/node_modules/.pnpm/@miniflare+core@2.12.1/node_modules/@miniflare/core/src/standards/event.ts:291:19)
    at node_modules/.pnpm/@prisma+client@4.11.0_prisma@4.11.0/node_modules/.prisma/client/edge.js (/workspace/repro-prisma-client-cloud-flare-pages/functions/[[path]].js:23896:67)
```

**Comment by inssein** [comment]
Oh man I can't remember anymore, I ended up just creating a completely separate API layer that my remix + cloudflare application was using. I know I couldn't fully get it working for all the scenarios I cared about, and moved away from the solution.

**Comment by janpio** [comment]
Totally understandable. This is nasty. We'll try to figure it out, step by step. Thanks for your inforamtion anyway.

**Comment by janpio** [comment]
A quick dump of thoughts based on an [internal discussion](https://prisma-company.slack.com/archives/C04FDERNP32/p1677789961015729) and some of my open tabs:

What is causing the problem and error message is indeed this code (assumes `url = env(FOO)` in the schema file):

```js
config.injectableEdgeEnv = {
  parsed: {
    FOO: typeof globalThis !== 'undefined' && globalThis['FOO'] || typeof process !== 'undefined' && process.env && process.env.FOO|| undefined
  }
}
```
generated in the generator for `edge.js` via https://github.com/prisma/prisma/blob/b70f3a52fedb9b36cac6031e2e56d13c8d70f1b9/packages/client/src/generation/utils/buildInjectableEdgeEnv.ts#L72-L79

Wrangler from Cloudflare does not like when you access a env var via `globalThis['FOO']` (but only if the env var actually is set and has a value, otherwise it does not care). 

Wrangler's motivation for this is probably that it used to be possible to access env vars via `globalThis` before, but is not any more now. So it defaults to warn or error for people that still do that - although when deploying to CF Pages itself this actually works just fine! It just is an empty value instead of the expected environment variable.

The code above is executed when you import Prisma Client right now. That way it can not know that you might have provided a different value for `datasources.db.url` in the `PrismaClient` constructor call.

Even if we would make it a getter, the way it is called right now does not take the presence of `datasources.db.url` into account and would execute it anyway. It is also unclear if that would really be possible as there are quite a few variants.

We could probably protect this code with some check if we are on CloudFlare somehow (https://developers.cloudflare.com/pages/platform/build-configuration/#environment-variables maybe?), to avoid calling `globalThis['FOO']`. 

That would solve the error message (good) and avoid having to set an invalid `env()` in the Prisma schema - but one would still need to supply the actual value via manual code via the PrismaCLient constructor. Optimally we want to avoid that.

**Comment by janpio** [comment]
Opened a discussion around this in the Cloudflare community: https://community.cloudflare.com/t/how-to-access-environment-variables-from-libraries/480041 (includes some additonal context on our current understanding what is going on)

**Comment by juliesaia** [comment]
Somewhat related, I tried following the steps for setting this up on Vercel Edge Functions (https://www.prisma.io/blog/database-access-on-the-edge-8F0t1s1BqOJE) but I'm getting the same issue, my environment variables aren't being set at all during runtime. Setting it with `process.env.DATABASE_URL` in the constructor isn't working either.

**Comment by janpio** [comment]
Can you open a new issue @juliesaia? Optimally describe a bit what your function looks like (if different than the blog post etc) and we will try to reproduce. Very possible this is related - but still better to have in a separate issue so we can discuss it. THanks!

**Comment by Ehesp** [comment]
Just came across this one... I ended up using a CF system environment variable in my schema file, and then overriding the constructor options:

```
datasource db {
  provider = "mongodb"
  url      = env("CLOUDFLARE_ACCOUNT_ID")
}
```

```js
let db: PrismaClient;

declare global {
  var __db: PrismaClient | undefined;
}

export function getDb(args: DataFunctionArgs) {
  const options = {
    datasources: {
      db: {
        url: args.context.DATABASE_URL as string,
      }
    }
  };

  if (process.env.NODE_ENV === "production") {
    db = new PrismaClient(options);
  } else {
    if (!global.__db) {
      global.__db = new PrismaClient(options);
    }
    db = global.__db!;
  }

  return db;
}
```

Seems to be working - it's not moaning since `CLOUDFLARE_ACCOUNT_ID` is known to Wrangler I guess.

**Comment by Ehesp** [comment]
Although now I'm running into https://github.com/prisma/prisma/issues/18104 😅

**Comment by nicognaW** [comment]
Hi, I'm having the same problem trying to access the Prisma data proxy from Cloudflare Workers, and I've seen a [reply](https://community.cloudflare.com/t/how-to-access-environment-variables-from-libraries/480041/2) from the Cloudflare community. Prisma edge client is not stable in Cloudflare Workers due to their new environment variable changes.

In my case, my Prisma schema was shared in multiple apps and the workaround mentioned above did not work in my project.

Is there a nightly version that solves this problem that I can use temporarily?

**Comment by uicrafts** [comment]
This seems to still be an issue today with the latest version of Prisma.

In my case, I have a monorepo with a `db` package that creates the Prisma client (using `'@prisma/client/edge'`). Another package in the same repo imports the client, and when starting up the Cloudflare Worker server via Wrangler, I get this error: `InvalidDatasourceError: Datasource "db" references an environment variable`

@janpio what's the recommended solution?

**Comment by millsp** [comment]
I have tried this with two reproductions and can confirm this is no longer an issue on recent wrangler versions. As for the confusing error messages regarding the environment variables, we have just improved it.

**Comment by dev-rajm** [comment]
I fixed this problem by rename the `.env` to `.dev.vars` and then restart the server.
